reconcile/ocm_upgrade_scheduler.py

```diff
@@ -138,45 +140,93 @@ def get_version_history(dry_run, upgrade_policies, ocm_map):
+
+    Args:
+        version (string): version to check
+        history (dict): history of versions per OCM instance
+        ocm_name (string): name of OCM instance
+        upgrade_conditions (dict): query results of upgrade conditions
+        workloads (list): strings representing types of workloads
+
+    Returns:
+        bool: are version upgrade conditions met
+    """
+    conditions_met = True
+    # check soak days condition is met for this version
+    soak_days = upgrade_conditions.get('soakDays', None)
+    if soak_days is not None:
+        ocm_history = history[ocm_name]
+        version_history = ocm_history['versions'].get(version, {})
+        for w in workloads:
+            workload_history = version_history.get('workloads', {}).get(w, {})
+            if soak_days > workload_history.get('soak_days', 0.0):
+                conditions_met = False
+
+    return conditions_met
+
 
-    for c in current_state:
-        d = [d for d in desired_state
-             if d.items() <= c.items()]
-        if not d:
-            c['action'] = 'delete'
-            diffs.append(c)
+def calculate_diff(current_state, desired_state, ocm_map, version_history):
+    """Check available upgrades for each cluster in the desired state
+    according to upgrade conditions
 
-    return diffs, err
+    Args:
+        current_state (list): current state of upgrade policies
+        desired_state (list): desired state of upgrade policies
+        ocm_map (OCMMap): OCM clients per OCM instance
+        version_history (dict): version history per OCM instance
+
+    Returns:
+        list: upgrade policies to be applied
+    """
+    diffs = []
 
+    now = datetime.utcnow()
+    for d in desired_state:
+        # ignore clusters with an existing upgrade policy
+        cluster = d['cluster']
+        c = [c for c in current_state if c['cluster'] == cluster]
+        if c:
+            continue
+
+        # ignore clusters with an upgrade schedule not within the next 2 hours
+        schedule = d['schedule']
+        next_schedule = croniter(schedule).get_next(datetime)
+        next_schedule_in_hours = \
+            (next_schedule - now).total_seconds() / 3600  # seconds in hour
+        if next_schedule_in_hours > 2:
+            continue
+
+        ocm = ocm_map.get(cluster)
+        available_upgrades = \
+            ocm.get_available_upgrades(d['current_version'], d['channel'])
+        for version in reversed(available_upgrades):
+            if version_conditions_met(version, version_history, ocm.name,
+                                      d['workloads'], d['conditions']):
+                item = {
+                    'action': 'create',
+                    'cluster': cluster,
+                    'version': version,
+                    'schedule_type': 'manual',
+                    'next_run': next_schedule.strftime('%Y-%m-%dT%H:%M:%SZ'),
+                }
+                diffs.append(item)
+                break
```

Comment:
Due to my confusion [here](https://github.com/app-sre/qontract-reconcile/pull/1795/files#diff-035de9609235cce528e8b2280f5b2d18e688e611ac923346eb60c8ff2c5f9f77R596-R597), it's unclear to me if this is trying to upgrade to the newest or oldest available version.

Reply:
we're trying to upgrade to the latest version